File: pkg/lexer/lexer.go
package lexer

import (
	"strings"
	"unicode"
)

// TokenType represents the type of a SQL token
type TokenType int

const (
	TokenError TokenType = iota
	TokenEOF
	TokenKeyword
	TokenIdentifier
	TokenString
	TokenNumber
	TokenSymbol
	TokenComment
	TokenWhitespace
)

// Token represents a Lexical token
type Token struct {
	Type  TokenType
	Value string
	Pos   int
}

var keywords = map[string]bool{
	"SELECT": true, "FROM": true, "WHERE": true, "INSERT": true, "INTO": true,
	"UPDATE": true, "DELETE": true, "CREATE": true, "DROP": true, "ALTER": true,
	"TABLE": true, "INDEX": true, "VIEW": true, "SEQUENCE": true, "SYNONYM": true,
	"JOIN": true, "LEFT": true, "RIGHT": true, "INNER": true, "OUTER": true,
	"ON": true, "AND": true, "OR": true, "AS": true, "GROUP": true, "BY": true,
	"ORDER": true, "HAVING": true, "DISTINCT": true, "UNION": true, "VALUES": true,
	"SET": true, "TRUNCATE": true, "GRANT": true, "REVOKE": true, "COMMIT": true, "ROLLBACK": true,
	"TRIGGER": true, "PROCEDURE": true, "FUNCTION": true, "PACKAGE": true, "BODY": true,
	"CONSTRAINT": true, "PRIMARY": true, "KEY": true, "FOREIGN": true, "REFERENCES": true,
	"CHECK": true, "UNIQUE": true, "DEFAULT": true, "NULL": true, "NOT": true, "MERGE": true,
}

// LexSQL lexes the input SQL string into a slice of tokens
func LexSQL(input string) []Token {
	var tokens []Token
	runes := []rune(input)
	length := len(runes)
	pos := 0

	for pos < length {
		r := runes[pos]

		// Whitespace
		if unicode.IsSpace(r) {
			start := pos
			for pos < length && unicode.IsSpace(runes[pos]) {
				pos++
			}
			tokens = append(tokens, Token{Type: TokenWhitespace, Value: string(runes[start:pos]), Pos: start})
			continue
		}

		// Comments
		if r == '-' && pos+1 < length && runes[pos+1] == '-' {
			start := pos
			for pos < length && runes[pos] != '\n' {
				pos++
			}
			tokens = append(tokens, Token{Type: TokenComment, Value: string(runes[start:pos]), Pos: start})
			continue
		}
		if r == '/' && pos+1 < length && runes[pos+1] == '*' {
			start := pos
			pos += 2
			for pos < length {
				if runes[pos] == '*' && pos+1 < length && runes[pos+1] == '/' {
					pos += 2
					break
				}
				pos++
			}
			tokens = append(tokens, Token{Type: TokenComment, Value: string(runes[start:pos]), Pos: start})
			continue
		}

		// Strings (Single quotes)
		if r == '\'' {
			start := pos
			pos++
			for pos < length {
				if runes[pos] == '\'' {
					if pos+1 < length && runes[pos+1] == '\'' {
						pos += 2 // Escaped quote
						continue
					}
					pos++
					break
				}
				pos++
			}
			tokens = append(tokens, Token{Type: TokenString, Value: string(runes[start:pos]), Pos: start})
			continue
		}

		// Identifiers (double quoted)
		if r == '"' {
			start := pos
			pos++
			for pos < length {
				if runes[pos] == '"' {
					pos++
					break
				}
				pos++
			}
			tokens = append(tokens, Token{Type: TokenIdentifier, Value: string(runes[start:pos]), Pos: start})
			continue
		}

		// Identifiers or Keywords (unquoted)
		if isIdentStart(r) {
			start := pos
			for pos < length && isIdentPart(runes[pos]) {
				pos++
			}
			value := string(runes[start:pos])
			typ := TokenIdentifier
			if isKeyword(value) {
				typ = TokenKeyword
			}
			tokens = append(tokens, Token{Type: typ, Value: value, Pos: start})
			continue
		}

		// Numbers
		if unicode.IsDigit(r) {
			start := pos
			for pos < length && (unicode.IsDigit(runes[pos]) || runes[pos] == '.') {
				pos++
			}
			tokens = append(tokens, Token{Type: TokenNumber, Value: string(runes[start:pos]), Pos: start})
			continue
		}

		// Symbols
		tokens = append(tokens, Token{Type: TokenSymbol, Value: string(r), Pos: pos})
		pos++
	}

	return tokens
}

func isIdentStart(r rune) bool {
	return unicode.IsLetter(r) || r == '_'
}

func isIdentPart(r rune) bool {
	return unicode.IsLetter(r) || unicode.IsDigit(r) || r == '_' || r == '$' || r == '#'
}

func isKeyword(s string) bool {
	return keywords[strings.ToUpper(s)]
}

File: crates/sql-processor-core/src/lexer.rs
use std::sync::LazyLock;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TokenKind {
    Error,
    Eof,
    Keyword,
    Identifier,
    String,
    Number,
    Symbol,
    Comment,
    Whitespace,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Token {
    pub kind: TokenKind,
    pub value: String,
    pub pos: usize,
}

static KEYWORDS: LazyLock<Vec<&'static str>> = LazyLock::new(|| {
    vec![
        "SELECT",
        "FROM",
        "WHERE",
        "INSERT",
        "INTO",
        "UPDATE",
        "DELETE",
        "CREATE",
        "DROP",
        "ALTER",
        "TABLE",
        "INDEX",
        "VIEW",
        "SEQUENCE",
        "SYNONYM",
        "JOIN",
        "LEFT",
        "RIGHT",
        "INNER",
        "OUTER",
        "ON",
        "AND",
        "OR",
        "AS",
        "GROUP",
        "BY",
        "ORDER",
        "HAVING",
        "DISTINCT",
        "UNION",
        "VALUES",
        "SET",
        "TRUNCATE",
        "GRANT",
        "REVOKE",
        "COMMIT",
        "ROLLBACK",
        "TRIGGER",
        "PROCEDURE",
        "FUNCTION",
        "PACKAGE",
        "BODY",
        "CONSTRAINT",
        "PRIMARY",
        "KEY",
        "FOREIGN",
        "REFERENCES",
        "CHECK",
        "UNIQUE",
        "DEFAULT",
        "NULL",
        "NOT",
        "MERGE",
    ]
});

pub fn lex_sql(input: &str) -> Vec<Token> {
    let mut tokens = Vec::new();
    let mut cursor = 0;
    while cursor < input.len() {
        let ch = input[cursor..]
            .chars()
            .next()
            .expect("cursor is a char boundary");
        let size = ch.len_utf8();

        if ch.is_whitespace() {
            let start = cursor;
            cursor += size;
            while cursor < input.len() {
                let next = input[cursor..].chars().next().expect("char boundary");
                if !next.is_whitespace() {
                    break;
                }
                cursor += next.len_utf8();
            }
            tokens.push(token(TokenKind::Whitespace, input, start, cursor));
            continue;
        }

        if input[cursor..].starts_with("--") {
            let start = cursor;
            cursor += 2;
            while cursor < input.len() && input.as_bytes()[cursor] != b'\n' {
                let next = input[cursor..].chars().next().expect("char boundary");
                cursor += next.len_utf8();
            }
            tokens.push(token(TokenKind::Comment, input, start, cursor));
            continue;
        }

        if input[cursor..].starts_with("/*") {
            let start = cursor;
            cursor += 2;
            while cursor < input.len() {
                if input[cursor..].starts_with("*/") {
                    cursor += 2;
                    break;
                }
                let next = input[cursor..].chars().next().expect("char boundary");
                cursor += next.len_utf8();
            }
            tokens.push(token(TokenKind::Comment, input, start, cursor));
            continue;
        }

        if matches!(ch, 'q' | 'Q')
            && input[cursor + size..].starts_with("'")
            && let Some(end) = scan_alternative_quote(input, cursor)
        {
            tokens.push(token(TokenKind::String, input, cursor, end));
            cursor = end;
            continue;
        }

        if ch == '\'' {
            let start = cursor;
            cursor += 1;
            while cursor < input.len() {
                if input.as_bytes()[cursor] == b'\'' {
                    if input[cursor + 1..].starts_with("'") {
                        cursor += 2;
                        continue;
                    }
                    cursor += 1;
                    break;
                }
                let next = input[cursor..].chars().next().expect("char boundary");
                cursor += next.len_utf8();
            }
            tokens.push(token(TokenKind::String, input, start, cursor));
            continue;
        }

        if ch == '"' {
            let start = cursor;
            cursor += 1;
            while cursor < input.len() {
                let next = input[cursor..].chars().next().expect("char boundary");
                cursor += next.len_utf8();
                if next == '"' {
                    break;
                }
            }
            tokens.push(token(TokenKind::Identifier, input, start, cursor));
            continue;
        }

        if is_ident_start(ch) {
            let start = cursor;
            cursor += size;
            while cursor < input.len() {
                let next = input[cursor..].chars().next().expect("char boundary");
                if !is_ident_part(next) {
                    break;
                }
                cursor += next.len_utf8();
            }
            let value = &input[start..cursor];
            let kind = if is_keyword(value) {
                TokenKind::Keyword
            } else {
                TokenKind::Identifier
            };
            tokens.push(Token {
                kind,
                value: value.to_string(),
                pos: start,
            });
            continue;
        }

        if ch.is_ascii_digit() {
            let start = cursor;
            cursor += size;
            while cursor < input.len() {
                let next = input[cursor..].chars().next().expect("char boundary");
                if !next.is_ascii_digit() && next != '.' {
                    break;
                }
                cursor += next.len_utf8();
            }
            tokens.push(token(TokenKind::Number, input, start, cursor));
            continue;
        }

        tokens.push(token(TokenKind::Symbol, input, cursor, cursor + size));
        cursor += size;
    }
    tokens
}

fn token(kind: TokenKind, input: &str, start: usize, end: usize) -> Token {
    Token {
        kind,
        value: input[start..end].to_string(),
        pos: start,
    }
}

fn is_ident_start(ch: char) -> bool {
    ch.is_alphabetic() || ch == '_'
}
fn is_ident_part(ch: char) -> bool {
    ch.is_alphanumeric() || matches!(ch, '_' | '$' | '#')
}
fn is_keyword(value: &str) -> bool {
    KEYWORDS
        .iter()
        .any(|keyword| keyword.eq_ignore_ascii_case(value))
}

fn scan_alternative_quote(input: &str, start: usize) -> Option<usize> {
    let mut cursor = start + input[start..].chars().next()?.len_utf8();
    if !input[cursor..].starts_with("'") {
        return None;
    }
    cursor += 1;
    let open = input[cursor..].chars().next()?;
    if open.is_whitespace() {
        return None;
    }
    cursor += open.len_utf8();
    let close = matching_quote_delimiter(open);
    while cursor < input.len() {
        let next = input[cursor..].chars().next()?;
        if next == close {
            let after = cursor + next.len_utf8();
            if input[after..].starts_with("'") {
                return Some(after + 1);
            }
        }
        cursor += next.len_utf8();
    }
    None
}

pub fn matching_quote_delimiter(open: char) -> char {
    match open {
        '[' => ']',
        '{' => '}',
        '(' => ')',
        '<' => '>',
        other => other,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn keeps_oracle_alternative_quoted_strings_together() {
        let tokens = lex_sql("select q'[first; second;]' from dual;");
        assert!(
            tokens
                .iter()
                .any(|token| token.kind == TokenKind::String && token.value.contains("first;"))
        );
    }
}
